ONSdigital/florence
# First time sign in screens

### What

First time Sign in screens all linked up

Refactor the forgotten password components slightly to reduce duplication. Large parts of it are used for the First time Sign in change password.

In addition have given the Sign in screen a bit of a refactor, aware that there are further improvements that can be done on this screen mind.

The user journeys for Authentication should be similar to the initial designs:
https://www.figma.com/file/TykOuMm6oKChMHw0pbmlkx/Sign-In?node-id=31%3A0

### How to review

- Check that the code looks ok
- Give the new feature a spin ( First time sign in )
- Check sign in still works
- Check forgotten password journey still works

#### Check First time Sign in

To check fully there are a few steps that need to be jumped through, but these guided steps should give you all that you need.

If you already have a Cognito account for whichever user pool you are testing against and you wish to reuse that account, you can delete it via the AWS CLI command:
`aws cognito-idp admin-delete-user --user-pool-id {user-pool-id} --username {email-address}`

Startup the dp-identity-api by running: `make debug ALLOWED_EMAIL_DOMAINS="@ons.gov.uk,@ext.ons.gov.uk,@gmail.com"` Please note that you will need to use an email for a latter step that you have access to the inbox of, please replace the `ALLOWED_EMAIL_DOMAINS` with an email domain you will later use for this.

In Florence run `make node-modules` and then proceed to start Florence server-side with the feature flag `ENABLE_NEW_SIGN_IN=true` (Via the usual `make debug` command)

Then create a new account with a REST client (e.g. ARC or POSTMAN)

```
Endpoint: http://localhost:25600/v1/users
Method: POST
Body: {
  "forename": "something",
  "lastname": "somethingElse",
  "email": "youremail@gmail.com"
}
```

(Note you must be able to access whatever email inbox you are using for this test and you will need to ensure the domain is valid, matching one you used to start the application with in the first step).

Once complete checkout your email account for that user. You should have an invitation email. Inside that email is a link to Florence (it will likely be pointing to publishing in develop, there is work to set this up for local dev which is WIP). Copy the password code from the email (note if you triple-click it to select all there will be an additional space at the beginning which is also copied - there is a task to adjust the email template to adjust for this).
Go to `http://localhost:8081/florence/login` enter your email and paste or type the password code from the email, removing any white space which might have been copied at the beginning of the password.

You should be then told to change your password. Go ahead and do this. You will need a Symbol character for the time being included in your password - a bug ticket has been raised on the backend to fix this issue (just a config issue). When complete you should see a screen saying so. This screen has a button 'start using florence' it makes a request to the permissions endpoint but currently, this endpoint is not functioning as expected there is a backend ticket to resolve this issue.

#### Test Sign in still works

Open the browsers developer tools and open the network tab
With your new password and email attempt to sign in, this will error saying failed to get permission - there is work on the backend to resolve this. But, in the network tab ensure you see a request to the `/tokens` endpoint that is a success.

#### Test the forgotten password journey

On the sign-in screen click the forgotten password link
Enter your email
Go to your inbox
Click the link (note it will likely take you to develop publishing (this is fine, the email template is a bit WIP in another task and will later be using a lambda) copy everything in the URL from `/florence` onwards and then go back to `http://localhost:8081/florence/` and replace `/florence` with the contents of your clipboard. The URL now looks something like: `https://publishing.develop.onsdigital.co.uk/florence/password-reset?vid=123456&uid=1234a56b-f0dc-478d-b065-9bfdac0000f0`

Enter a new password, note like before this will require a symbol character in addition to the other validation rules listed (a change to the backend is planned to resolve this)

### Who can review

Anyone except me
